torch/testing/_internal/distributed/rpc/rpc_test.py

```diff
@@ -2731,6 +2731,49 @@ class TestPickler:
             torch.distributed.rpc.api._default_pickler is _internal_rpc_pickler
         )
 
+    @dist_init
+    def test_wait_all(self):
+        with rpc.wait_all():
+            self.assertTrue(rpc.thread_local_var.future_list == [])
+            dst = worker_name((self.rank + 1) % self.world_size)
+            fut = rpc.rpc_async(dst, torch.add, (torch.ones(2, 2), 1))
+            self.assertTrue(len(rpc.thread_local_var.future_list) == 1)
+            self.assertEqual(fut.wait(), torch.ones(2, 2) + 1)
+        self.assertFalse(hasattr(rpc.thread_local_var, "future_list"))
+
+    @dist_init
+    def test_wait_all_multiple_call(self):
+        with rpc.wait_all():
+            self.assertTrue(rpc.thread_local_var.future_list == [])
+            dst = worker_name((self.rank + 1) % self.world_size)
+            for i in range(20):
+                fut = rpc.rpc_async(dst, torch.add, (torch.ones(i, i), 1))
+                res = rpc.rpc_sync(dst, torch.add, (torch.ones(i, i), 1))
+                self.assertEqual(res, torch.ones(i, i) + 1)
+                self.assertEqual(fut.wait(), torch.ones(i, i) + 1)
+            self.assertTrue(len(rpc.thread_local_var.future_list) == 40)
+        self.assertFalse(hasattr(rpc.thread_local_var, "future_list"))
+
+    @dist_init
+    def test_wait_all_timeout(self):
+        expected_error = get_timeout_error_regex(dist_utils.TEST_CONFIG.rpc_backend_name)
+        with self.assertRaisesRegex(RuntimeError, expected_error):
+            with rpc.wait_all():
+                self.assertTrue(rpc.thread_local_var.future_list == [])
+                dst = worker_name((self.rank + 1) % self.world_size)
+                timeout = 0.1  # 100 ms
+                fut = rpc.rpc_async(dst, my_sleep_func, args=(1,), timeout=timeout)
+        self.assertTrue(len(rpc.thread_local_var.future_list) == 1)
+
+    @dist_init
+    def test_wait_all_raise_in_user_func(self):
+        with self.assertRaises(ValueError):
+            with rpc.wait_all():
+                self.assertTrue(rpc.thread_local_var.future_list == [])
+                dst = worker_name((self.rank + 1) % self.world_size)
+                fut = rpc.rpc_async(dst, raise_func)
+        self.assertTrue(len(rpc.thread_local_var.future_list) == 1)
```

Comment:
@mrshenli  Note if an exception is raised by any future in future_list that is waited on, and that exception is caught in the application, future_list will not be deleted.  Ie:

`try:
	with rpc.wait_all():
		fut = rpc.rpc_async(dst, torch.add, (torch.ones(2, 2), 1))
		fut = rpc.rpc_async(dst, raise_func)
except BaseException:
	#rpc.thread_local_var.future_list still exists with 2 futures in it`

Since future_list is reinitialized as [] each time wait_all Is invoked, this should be fine?  On the other hand we could catch any error raised by torch.futures.wait_all and delete future_list before passing the error up, but the user may want access to the future_list in the event of an exception.  What do you think?

Reply:
>  On the other hand we could catch any error raised by torch.futures.wait_all and delete future_list before passing the error up, but the user may want access to the future_list in the event of an exception. What do you think?

Good call! We can add a `collect_all()` API (similar to [the one in `torch.futures`](https://pytorch.org/docs/master/futures.html#torch.futures.collect_all)) for this purpose. I mean sth like:

```python
with rpc.collect_all() as futs:
   # make rpc calls

   # access futs here if necessary
```

For `wait_all()` I'd vote for letting it delete `future_list` and  re-throw the first exception. Open to other suggestions. :)

Reply:
We can add `collect_all()` in a followup PR.